### PR TITLE
BAU-559 - Replace custom guard with custom user provider

### PR DESCRIPTION
**Pull request by LeeBrooks3**

### CODE REVIEW

File: src/Auth/UserProvider.php [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        try {
+            $response = $this->client->getUserByToken('Bearer', $token);
+
+            $user = $response['data'];
+        } catch (\Exception $exception) {
+            return null;
+        }
+
+        return parent::retrieveByCredentials([
+            'email' => $user['email'], // TODO: update to use uuid / auth id
```
Comment (mikefrancis) [inline src/Auth/UserProvider.php]: I would maybe move the TODO to the description

File: src/ServiceProvider.php [hunk 2]
```diff
@@ -41,13 +40,18 @@ public function boot()
...
         ]);
 
-        $this->app['auth']->extend('passport', function (Application $app, $name, $config) {
-            $userProvider = $app['auth']->createUserProvider($config['provider']);
-            $session = $app->make(Session::class);
-            $request = $app->make(Request::class);
+        /** @var AuthManager $authManager */
+        $authManager = $this->app->make(AuthManager::class);
+
+        $authManager->provider('passport', function ($app, array $config) {
```
Comment (mikefrancis) [inline src/ServiceProvider.php]: I know how much you love type-hints, so you can type-hint `$app` with the `Application` contract here if you want

File: src/ServiceProvider.php [hunk 3]
```diff
@@ -41,13 +40,18 @@ public function boot()
             $this->getConfigPath() => config_path('passport.php'),
         ]);
 
-        $this->app['auth']->extend('passport', function (Application $app, $name, $config) {
-            $userProvider = $app['auth']->createUserProvider($config['provider']);
-            $session = $app->make(Session::class);
-            $request = $app->make(Request::class);
+        /** @var AuthManager $authManager */
+        $authManager = $this->app->make(AuthManager::class);
```
Comment (mikefrancis) [inline src/ServiceProvider.php]: Can DI this into the method instead if this cleans it up at all

File: tests/TestCase.php [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+    protected $container;
+
+    /**
+     * Sets up the tests.
+     */
+    public function setUp()
+    {
+        parent::setUp();
+
+        $this->container = Container::getInstance();
```
Comment (mikefrancis) [inline tests/TestCase.php]: What's this for?

Comment (LeeBrooks3) [reply tests/TestCase.php]: Ah I was going to start writing a test for the user provider but it was going to be pretty pointless.. can remove.

File: src/ServiceProvider.php (hunk 2 above)
Comment (LeeBrooks3) [reply src/ServiceProvider.php]: my man

File: src/Auth/UserProvider.php [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+     *
+     * @return bool
+     */
+    public function validateCredentials(Authenticatable $user, array $credentials)
+    {
+        $token = array_get($credentials, 'api_token');
+
+        if ($token) {
+            try {
+                $this->client->checkAccessToken($credentials['api_token']);
```
Comment (Tyrannogina) [inline src/Auth/UserProvider.php]: We could use `$token` here instead of reading from the array again.

**Pull request merged**